File: tsasm/arm/regen/neon_encode_test.pl
#! /usr/bin/env perl
# SPDX-License-Identifier: BSD-3-Clause
#
# Cross-checks the Perl NEON encoders in neon_encode.pl against
# arm-linux-gnueabihf-as. Reads NEON GAS lines on stdin (one per
# line); for each, computes the Perl encoding and the arm-as
# encoding and reports any mismatch.
#
# Usage:
#   perl neon_encode_test.pl <neon_lines.txt
#
# A typical run (use the Go wrapper for an end-to-end automated
# version, see TestNEONEncoderAgainstGAS in neon_encode_test.go):
#   perl neon_encode_test.pl < unique_neon_lines.txt
#
# Exits 0 only if every input line matched and the Perl encoder
# recognized every line (no skips). A skip means there's a missing
# encoder in neon_encode.pl, which the regen pipeline now treats as
# a hard error (batch_encode_neon dies rather than falling back to
# arm-as), so we want the same strict behavior here.

use strict;
use warnings;
use File::Temp qw/tempfile/;
use FindBin;
require "$FindBin::Bin/neon_encode.pl";

sub asm_encode_batch {
    my @lines = @_;
    my ($sf, $sp) = tempfile("nv-XXXX", SUFFIX => ".s", TMPDIR => 1);
    my ($of, $op) = tempfile("nv-XXXX", SUFFIX => ".o", TMPDIR => 1);
    my ($bf, $bp) = tempfile("nv-XXXX", SUFFIX => ".bin", TMPDIR => 1);
    close $of; close $bf;
    print $sf ".syntax unified\n.arch armv7-a\n.fpu neon\n.code 32\n";
    for my $l (@lines) { print $sf "$l\n"; }
    close $sf;
    system("arm-linux-gnueabihf-as", "-mfpu=neon", "-march=armv7-a",
           "-o", $op, $sp) == 0
        or die "arm-as failed; input at $sp\n";
    system("arm-linux-gnueabihf-objcopy", "-O", "binary",
           "--only-section=.text", $op, $bp) == 0
        or die "objcopy failed\n";
    open my $fh, "<:raw", $bp or die;
    my $buf;
    read $fh, $buf, -s $bp;
    close $fh;
    unlink $sp, $op, $bp;
    my @words = unpack("V*", $buf);
    return @words;
}

my @lines;
while (my $l = <STDIN>) {
    chomp $l;
    $l =~ s/^\s+|\s+$//g;
    next unless length $l;
    next if $l =~ /^[#@\.]/;
    push @lines, $l;
}

my @asm = asm_encode_batch(@lines);
if (@asm != @lines) {
    die sprintf("arm-as produced %d words for %d lines\n",
                scalar @asm, scalar @lines);
}

my $ok       = 0;
my $mismatch = 0;
my @skipped;
for (my $i = 0; $i < @lines; $i++) {
    my $line = $lines[$i];
    my $want = $asm[$i];
    my $got  = eval { encode_neon($line) };
    if ($@) {
        push @skipped, $line;
        next;
    }
    if ($got == $want) {
        $ok++;
    } else {
        $mismatch++;
        printf STDERR "MISMATCH: %-40s perl=0x%08x asm=0x%08x\n",
            $line, $got, $want;
    }
}

printf "ok=%d  mismatch=%d  skipped(no encoder)=%d\n",
    $ok, $mismatch, scalar @skipped;
for my $l (@skipped) {
    print STDERR "SKIPPED: $l\n";
}
exit(1) if $mismatch || @skipped;
exit(0);

File: tsasm/arm/regen/neon_encode_test.go
// SPDX-License-Identifier: BSD-3-Clause

package regen

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"testing"
)

// TestNEONEncoderAgainstGAS extracts every unique NEON GAS line from
// the cached upstream .S files and feeds them through
// neon_encode_test.pl, which encodes each line with both our pure-Perl
// encoder and arm-linux-gnueabihf-as and exits non-zero on any
// mismatch or unrecognized line. Together with TestRegenReproducible
// this gives end-to-end coverage of the regen pipeline.
//
// Gated identically to TestRegenReproducible: skipped by default,
// runs under CI=true or --run-regen-tests. Additionally needs
// arm-linux-gnueabihf-as / arm-linux-gnueabihf-objcopy (the cross-as
// is only used by this validator; the production regen pipeline no
// longer touches it).
func TestNEONEncoderAgainstGAS(t *testing.T) {
	if !*runRegenTests && os.Getenv("CI") != "true" {
		t.Skip("regen integration test is gated: pass --run-regen-tests or set CI=true")
	}
	if runtime.GOOS == "windows" {
		t.Skip("test pipeline assumes a Unix shell")
	}
	requireTool(t, "perl", "perl")
	requireTool(t, "arm-linux-gnueabihf-as", "binutils-arm-linux-gnueabihf")
	requireTool(t, "arm-linux-gnueabihf-objcopy", "binutils-arm-linux-gnueabihf")

	regenDir, err := filepath.Abs(".")
	if err != nil {
		t.Fatal(err)
	}

	// Make sure the .cache/ has the upstream .S files. Easiest way is
	// to run regen.sh once; on first run that fetches the upstream .pl
	// from the network. Subsequent runs use the cache.
	if _, err := os.Stat(filepath.Join(regenDir, ".cache", "chacha-armv4.S")); err != nil {
		cmd := exec.Command("sh", filepath.Join(regenDir, "regen.sh"))
		cmd.Dir = regenDir
		cmd.Stderr = os.Stderr
		cmd.Stdout = os.Stderr
		if err := cmd.Run(); err != nil {
			t.Fatalf("priming regen.sh: %v", err)
		}
	}

	lines, err := uniqueNEONLines(
		filepath.Join(regenDir, ".cache", "chacha-armv4.S"),
		filepath.Join(regenDir, ".cache", "poly1305-armv4.S"),
	)
	if err != nil {
		t.Fatalf("extracting NEON lines: %v", err)
	}
	if len(lines) == 0 {
		t.Fatal("no NEON lines found in cached .S files")
	}
	t.Logf("checking %d unique NEON instruction forms", len(lines))

	cmd := exec.Command("perl", filepath.Join(regenDir, "neon_encode_test.pl"))
	cmd.Dir = regenDir
	cmd.Stdin = strings.NewReader(strings.Join(lines, "\n") + "\n")
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &out
	if err := cmd.Run(); err != nil {
		t.Fatalf("neon_encode_test.pl failed:\n%s", out.String())
	}
	t.Logf("neon_encode_test.pl: %s", strings.TrimSpace(out.String()))
}

// uniqueNEONLines returns the sorted, deduplicated set of NEON GAS
// lines (those starting with "v[a-z]") across the given .S files.
func uniqueNEONLines(paths ...string) ([]string, error) {
	neonRE := regexp.MustCompile(`^v[a-z][a-z0-9.]*`)
	seen := map[string]bool{}
	for _, p := range paths {
		f, err := os.Open(p)
		if err != nil {
			return nil, fmt.Errorf("open %s: %w", p, err)
		}
		sc := bufio.NewScanner(f)
		for sc.Scan() {
			line := strings.TrimSpace(sc.Text())
			if i := strings.Index(line, "@"); i >= 0 {
				line = strings.TrimSpace(line[:i])
			}
			if !neonRE.MatchString(line) {
				continue
			}
			seen[line] = true
		}
		f.Close()
		if err := sc.Err(); err != nil {
			return nil, err
		}
	}
	out := make([]string, 0, len(seen))
	for l := range seen {
		out = append(out, l)
	}
	sort.Strings(out)
	return out, nil
}
